### PR TITLE
api for third party

### PR DESCRIPTION
**Pull request by nic-chen**

### CODE REVIEW

**Comment by moonming** [comment]
@nic-chen can you add some examples?

**Comment by moonming** [comment]
luackeck failed, please have a look:
```
luacheck -q lua
Checking lua/apisix/api/config.lua                6 warnings
    lua/apisix/api/config.lua:1:7: unused variable core
    lua/apisix/api/config.lua:2:7: unused variable routes
    lua/apisix/api/config.lua:3:7: unused variable schema_plugin
    lua/apisix/api/config.lua:4:7: unused variable tostring
    lua/apisix/api/config.lua:5:7: unused variable ipairs
    lua/apisix/api/config.lua:6:7: unused variable tonumber
Checking lua/apisix/api/init.lua                  5 warnings
    lua/apisix/api/init.lua:17:16: unused function check_signature
    lua/apisix/api/init.lua:31:11: unused variable str_base64
    lua/apisix/api/init.lua:32:32: accessing undefined variable str
    lua/apisix/api/init.lua:72:15: unused variable err
    lua/apisix/api/init.lua:72:21: accessing undefined variable check_token
```

**Comment by nic-chen** [comment]
@moonming  it's not a full version..
just for you guys to confirm:
1. is the way to expose api  ok 
2. is the code style ok ?

File: bin/apisix [hunk 1]
```diff
@@ -152,6 +152,21 @@ http {
     }
     {% end %}
 
+    {% if enable_api and port_api then %}
```
Comment (membphis) [inline bin/apisix]: only use `enable_api` is enough.

File: conf/config.yaml [hunk 2]
```diff
@@ -10,6 +10,11 @@ apisix:
     - 'unix:'
   # port_admin: 9180              # use a separate port
 
+  enable_api: true
+  port_api: 9988
+  allow_api: "127.0.0.0/24"  
+  api_key: "key4@pi6"
```
Comment (membphis) [inline conf/config.yaml]: we can support `api_key` later.

File: lua/apisix/api/init.lua [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        if not data then
+            core.log.error("invalid request body: ", req_body, " err: ", err)
+            core.response.exit(400, {error_msg = "invalid request body",
+                                     req_body = req_body})
+        end
+
+        req_body = data
+    end
+
+    --校验接口api_token
```
Comment (membphis) [inline lua/apisix/api/init.lua]: pls use English comment

File: lua/apisix.lua [hunk 4]
```diff
@@ -233,6 +235,18 @@ function _M.http_admin()
     end
 end
 
+function _M.http_api()
+    if not router then
+        router = api_init.get()
+    end
+
+    -- core.log.info("uri: ", get_var("uri"), " method: ", get_method())
```
Comment (membphis) [inline lua/apisix.lua]: Don't have to comment this line of code

File: lua/apisix/api/config.lua [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	--todo check hash to confirm config is changed
+	--todo parse config
+	
+	--todo load config into cache
+	--todo create/raise events
+    return 200, "ok"
+end
+
+
+return _M
```
Comment (membphis) [inline lua/apisix/api/config.lua]: The last line is preferably a blank line.

File: lua/apisix/api/init.lua [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+local _M = {version = 0.1}
+local router
+
+
+local function check_signature(signature, timestamp, res)
+    if not signature then
+        return nil, "api signature is needed"
+    end
+
+    local now = os.time()
```
Comment (membphis) [inline lua/apisix/api/init.lua]: We always prefer to use `ngx. *` function

File: lua/apisix/api/init.lua [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    end
+
+    local local_conf = core.config.local_conf()
+
+    local src        = res..'|'..timestamp
+    local digest     = ngx.hmac_sha1(local_conf.apisix.api_key, src)
+    local str_base64 = ngx.encode_base64(digest)
+    local generated  = ngx.md5(str_base64)
+
+    if generated==signature then
```
Comment (membphis) [inline lua/apisix/api/init.lua]: `if generated == signature then`

File: lua/apisix/api/init.lua [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        core.response.exit(404)
+    end
+
+    local method = str_lower(get_method())
+    if not resource[method] then
+        core.response.exit(404)
+    end
+
+    ngx.req.read_body()
+    local req_body = ngx.req.get_body_data()
```
Comment (membphis) [inline lua/apisix/api/init.lua]: https://github.com/openresty/lua-nginx-module#ngxreqget_body_data

This function returns nil if

* the request body has not been read,
* the request body has been read into disk temporary files,
* or the request body has zero size.

File: lua/apisix/api/init.lua [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            core.log.error("invalid request body: ", req_body, " err: ", err)
+            core.response.exit(400, {error_msg = "invalid request body",
+                                     req_body = req_body})
+        end
+
+        req_body = data
+    end
+
+    --校验接口api_token
+    local signature = req_body and req_body.signature or nil
```
Comment (membphis) [inline lua/apisix/api/init.lua]: remove `or nil`, it has no effect

**Pull request closed without merging**